### PR TITLE
Dc-Sync: add transformations support

### PR DESCRIPTION
**Pull request by e-orz**
Still without support for crashes and restarts. It will be with another PR to keep them small and concrete.

### CODE REVIEW

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 1]
```diff
@@ -847,20 +851,17 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
...
 
   def runSyncingEngine(dcInfo: DcInfo): SyncerMaterialization = {
+    val transformationsStr = if (dcInfo.transformations.nonEmpty) s" with transformations: ${dcInfo.transformations.mkString("[", ",", "]")}" else ""
     logger.info(
-      s"Starting sync engine for data center id ${dcInfo.id} from location ${dcInfo.location}${dcInfo.positionKey
-        .fold("")(key => s" using position key $key")}${dcInfo.tsvFile
-        .fold("")(f => " and file " + f)}"
+      s"Starting sync engine for data center id ${dcInfo.id} from location " +
+        s"${dcInfo.location}${dcInfo.positionKey.fold("")(key => s" using position key $key")}${dcInfo.tsvFile.fold("")(f => " and file " + f)}" +
+        s"$transformationsStr"
```
Comment (MoriaYechiel) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: Don't you want to write something like: 
s"**transformations are:** $transformationsStr"

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/Util.scala [hunk 2]
```diff
@@ -97,4 +98,36 @@ object Util extends LazyLogging {
...
+      val transformations = dcInfo.transformations.toList
+      infotonData => {
+        val oldMeta = infotonData.meta
+        val newMeta = oldMeta.copy(path = transform(transformations, oldMeta.path))
+        val infotonQuads = infotonData.data.utf8String.split('\n')
+        val newData = infotonQuads.foldLeft(StringBuilder.newBuilder) { (total, line) =>
+          val subjectEndPos = line.indexOf(' ')
+          val predicateEndPos = line.indexOf(' ', subjectEndPos + 1)
+          val isValueReference = line.charAt(predicateEndPos + 1) == '<'
+          val lastSpaceBeforeLastPart = line.lastIndexOf(' ', line.length - 3)
```
Comment (MoriaYechiel) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/Util.scala]: Why length - 3 ?

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala (hunk 1 above)
Comment (e-orz) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: It's already done. See line 854.

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/Util.scala [hunk 3]
```diff
@@ -97,4 +98,36 @@ object Util extends LazyLogging {
...
+          val isLastReference = line.charAt(line.length - 3) == '>' && !line.substring(lastSpaceBeforeLastPart).contains("^^")
+          val isQuad = isLastReference && lastSpaceBeforeLastPart != predicateEndPos
+          val valueEndPos = if (isQuad) lastSpaceBeforeLastPart else line.length - 2
+          val newSubject = transform(transformations, line.substring(0, subjectEndPos + 1))
+          val predicate = line.substring(subjectEndPos + 1, predicateEndPos + 1)
+          val oldValue = line.substring(predicateEndPos + 1, valueEndPos + 1)
+          val newValue = if (isValueReference) transform(transformations, oldValue) else oldValue
+          val newQuad = if (isQuad) transform(transformations, line.substring(valueEndPos + 1)) else line.substring(valueEndPos + 1)
```
Comment (MoriaYechiel) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/Util.scala]: Why is that needed? 
If not quad -> it's a value and you already took care of it
